lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs

```diff
@@ -0,0 +1,192 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Timers;
+using Org.Apache.REEF.Driver;
+using Org.Apache.REEF.Driver.Bridge.Events;
+using Org.Apache.REEF.IMRU.API;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.IMRU.OnREEF.Driver
+{
+    /// <summary>
+    /// JobLifecycleManager orchestrates job cancellation flow.
+    /// If job cancellation detector is configured in job definition, the manager starts a timer and periodically checks for cancellation signal.
+    /// if cancellation signal is detected, the manager creates JobCancelled event and propagates the event to all subscriber.
+    /// the manager is used by IMRU driver to enable job cancellation based on job definition.
+    /// </summary>
+    public class JobLifeCycleManager :
```

Comment:
yes, if we keep this as pull implementation
I added JobLifecycleManager class which essentially pulls cancellation signal from IJobCancellationDetector on timer.
Main reason I added it into driver project because, (imho) most of the clients will use external dependencies to detect cancellation and will want to implement some kind of pulling anyway, so this class makes implementation easier on the client side. Also this timer-based lifecycle manager could be useful to implement other scenarios in driver. For instance: job suspend/resume, stage timeouts ((hence generic name for the class)).
if we want keep driver configuration to be purely event driven, this implementation can be moved to the client: we can expose Observable<IJobCancelled> in the job definition and have client do the pulling implementation.
Thoughts?

---
In reply to: [84164147](https://github.com/apache/reef/pull/1161#discussion_r84164147) [](ancestors = 84164147)